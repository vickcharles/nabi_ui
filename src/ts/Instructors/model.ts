export interface ExperienceType {
  id?: number;
  employer: string;
  jobTitle: string;
  fromMonth: string;
  fromYear: string;
  toMonth: string;
  toYear: string;
  location: string;
}

export enum DegreeType {
  associate = 'Associate Degree',
  bachelors = 'Bachelor\'s Degree',
  graduate = 'Graduate Degreee',
  professional = 'Professional Degree',
  certification = 'Certification',
  other = 'Other'
}

export interface EducationType {
  id?: number;
  schoolName: string;
  graduationYear: string;
  degreeType: DegreeType;
  fieldOfStudy: string;
}

export enum PlaceForLessonsOptions {
  home = 'Student\'s home',
  studio = 'Instructor\'s studio',
  online = 'Online',
}

export interface PlaceForLessonsType {
  home: boolean;
  studio: boolean;
  online: boolean;
}

export enum LessonDuration {
  thirtyMins = '30 mins',
  fortyFiveMins = '45 mins',
  sixtyMins = '60 mins',
  ninetyMins = '90 mins'
}

export interface RatesType {
  thirtyMinsRate: number;
  fortyFiveMinsRate: number;
  sixtyMinsRate: number;
  ninetyMinsRate: number;
}

export enum SkillLevel {
  beginner = 'Beginner',
  intermediate = 'Intermediate',
  advanced = 'Advanced'
}

export interface InstrumentsType {
  instrument: string;
  skillLevel: SkillLevel;
}

export interface InstructorType {
  userId: string;
  bio?: string;
  instruments?: InstrumentsType[];
  rates?: RatesType;
  placeForLessons?: PlaceForLessonsType;
  studioAddress?: string;
  education?: EducationType[];
  experience?: ExperienceType[];
}

export type IState = InstructorType[];
