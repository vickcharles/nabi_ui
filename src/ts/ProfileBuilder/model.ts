export interface InstructorState {
  userId: string;
  bio?: string;
}

export type IState = InstructorState[];
  