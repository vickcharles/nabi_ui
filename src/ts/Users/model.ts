export enum HearAboutUs {
  Google = 'google',
  Facebook = 'facebook',
  Craigslist = 'craigslist',
  PrintAds = 'printAds',
  WordOfMouth = 'wordOfMouth',
  CurrentNabiUser = 'currentNabiUser',
  Other= 'other'
}

export interface HearAboutUsInfo {
  value: HearAboutUs;
  label: string;
}

export const hearAboutUsInfo = {
  Google: <HearAboutUsInfo> {
    value: HearAboutUs.Google,
    label: 'Google'
  },
  Facebook: <HearAboutUsInfo> {
    value: HearAboutUs.Facebook,
    label: 'Facebook'
  },
  Craigslist: <HearAboutUsInfo> {
    value: HearAboutUs.Craigslist,
    label: 'Craigslist'
  },
  PrintAds: <HearAboutUsInfo> {
    value: HearAboutUs.PrintAds,
    label: 'Print Ads'
  },
  WordOfMouth: <HearAboutUsInfo> {
    value: HearAboutUs.WordOfMouth,
    label: 'Word Of Mouth'
  },
  CurrentNabiUser: <HearAboutUsInfo> {
    value: HearAboutUs.CurrentNabiUser,
    label: 'Current Nabi User'
  },
  Other: <HearAboutUsInfo> {
    value: HearAboutUs.Other,
    label: 'Other'
  }
};

export enum Role {
  student = 'student',
  instructor = 'instructor',
  parent = 'parent'
}

export interface UserState {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  password: string;
  zipCode: string;
  role: Role;
  hearAboutUs: string;
  avatar?: string;
  displayName: string;
}

export type IState = UserState[];
