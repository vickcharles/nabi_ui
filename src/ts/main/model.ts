/**
 * State for firing redirect
 * @interface RedirectState
 */
export interface RedirectState {
  fireRedirect: boolean;
}
