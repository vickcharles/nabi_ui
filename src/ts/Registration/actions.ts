import { Action } from 'redux';
import { RegistrationActions } from './constants/ActionTypes';
import { UserState } from './model';

interface CreateUser extends Action {
  user: UserState;
}

interface ChangeAvatar extends Action {
  id: string;
  avatar: string;
}

// Create user
export function createUser(user: UserState): CreateUser {
  return {
    user,
    type: RegistrationActions.CREATE_USER
  };
}

/**
 * Action to change avatar
 * 
 * @param {string} id - The id of the user
 * @param {string} avatar - The avatar of the user
 */
export function changeAvatar(id: string, avatar: string): ChangeAvatar {
  return {
    id: id,
    avatar: avatar ,
    type: RegistrationActions.CHANGE_AVATAR
  };
}
