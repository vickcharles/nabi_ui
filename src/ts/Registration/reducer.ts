import { AnyAction } from 'redux';
import { UserState, IState } from './model';
import { RegistrationActions } from './constants/ActionTypes';
import { ProfileBuilderActions } from '../ProfileBuilder/constants/ActionTypes';

const initialState: UserState[] = [];

export default function usersReducer(state: IState = initialState, action: AnyAction): IState {
  switch (action.type) {
    case RegistrationActions.CREATE_USER:
      return [...state, action.user];
    
    case ProfileBuilderActions.CHANGE_BIO:
      return <IState> state.map(user =>
        user.id === action.user.id
          ? { 
            ...user, 
            ...action.user.bio,
            bio: action.user.bio
          }
          : user
      );

  //   case ProfileBuilderActions.CHANGE_BIO:
  //      if (state.length === 0) {
  //        let nuser: UserState = Object.assign(
  //          {},
  //          {
  //            firstName: '',
  //            lastName: '',
  //            email: '',
  //            password: '',
  //            zipCode: '',
  //            role: Role.instructor,
  //            hearAboutUs: '',
  //            avatar: ''},
  //          {avatar: action.avatar}
  //        );
  //        return [...state, nuser];
  //      }
  //      let newState = state.map((obj, i) => {
  //        if (obj.email === action.email) {
  //          obj.avatar = action.avatar;
  //        }
  //        return obj;
  //      });
  //      return [...newState];

    default:
      return state;
  }
}
