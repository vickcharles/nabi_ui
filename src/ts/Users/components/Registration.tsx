import * as React from 'react';
import { Redirect } from 'react-router-dom';
import * as moment from 'moment';

import RegistrationForm from './RegistrationForm';
import { UserState, Role } from '../model';
import { InstructorType } from '../../Instructors/model';
import { PageTitle } from '../../main';
import { RedirectState } from '../../main/model';
import AgeDisclaimer from './AgeDisclaimer';

import { fields, 
         validators, 
         validateFields, 
         RegistrationValidatorState, 
         ValidatorsInterface } from './RegistrationValidator';
/**
 * Props for registration
 * @interface RegistrationProps
 */
interface RegistrationProps {
  createUser: (user: UserState) => void;
  createInstructor: (instructor: InstructorType) => void;
  searchZipCode: (user: UserState) => void;
  role?: Role;
}

interface OpenModalState {
  openModal: boolean;
}

/**
 * State for Registration
 * @interface RegistrationState
 */
interface RegistrationState extends
UserState,
RedirectState,
RegistrationValidatorState,
OpenModalState {}

/**
 * Contains a form to register new users
 * @class Registration
 * @extends React.Component<RegistrationProps, UserState & RedirectState>
 */
export class Registration extends React.Component<RegistrationProps, RegistrationState> {
private validators: Array<ValidatorsInterface>;
  constructor(props: RegistrationProps) {
    super(props);
    this.state = {
      id: '',
      firstName: '',
      lastName: '',
      email: '',
      password: '',
      zipCode: '',
      role: Role.student,
      hearAboutUs: '',
      birthday: moment(),
      fireRedirect: false,
      displayName: '',
      openModal: false,
      fields: fields 
    };

    this.validators = validators;

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleBirthdayChange = this.handleBirthdayChange.bind(this);
  }

  componentWillMount() {
    if (this.props.role) {
        this.setState({role: this.props.role});
    }
  }

  public handleChange(event: any): void {
    const target = event.target;
    const value = target.value;
    const name = target.name;

    let newFieldsValue = Object.assign({}, this.state.fields);
    if ( this.state.fields[name] && true ) {
        newFieldsValue[name].touched = true;
    }

    this.setState({
      ...this.state,
      [name]: value,
      fields: newFieldsValue
    });
  }

  public handleBirthdayChange(date: moment.Moment): void {
    this.setState({ birthday: date });
  }

  public displayAgeDisclaimer(): void {
    if (this.state.birthday) {
      const userAge = Math.abs(this.state.birthday.diff(moment(), 'years'));

      if (userAge <= 17) {
        this.setState({ openModal: true });
      } else {
        if (!this.state.fireRedirect && !this.state.openModal) {
          this.setState({ fireRedirect: true }, () => {
            this.createUser();
          });
        }
      }
    }
  }

  public handleSubmit(event: React.SyntheticEvent<HTMLInputElement>): void {

    if (event) {
      event.preventDefault();
    }

    let temp = validateFields( this.state, this.validators);
    
    if ( temp.hasErrors && true) {
      this.setState({fields: temp.newValidateFields});
    } else {
      this.displayAgeDisclaimer();
      this.setState({ id: this.generateId() });

      if (!this.state.fireRedirect) {
        this.setState({ fireRedirect: true }, () => {
          this.createUser();
        });
      }
    }
  }
  
  public createUser(): void {
    if (this.state.fireRedirect === false) {
      this.setState({ fireRedirect: true });
    }

    const userValues: UserState = {
      id: this.state.id,
      firstName: this.state.firstName,
      lastName: this.state.lastName,
      email: this.state.email,
      password: this.state.password,
      zipCode: this.state.zipCode,
      role: this.state.role,
      hearAboutUs: this.state.hearAboutUs,
      displayName: `${this.state.firstName} ${this.state.lastName}`
    };

    this.props.createUser(userValues);

    const userId = {
      userId: this.state.id,
    };

    if ( this.state.role === Role.instructor) {
      this.props.createInstructor(userId);
    }

    if ( this.state.zipCode && true) {
      this.props.searchZipCode(userValues);
    }
  }

  /* The userID should be generated by the api
  TODO: remove this method when the api is hooked */
  public generateId(): string {
    return `uniqueid${new Date().getUTCMilliseconds()}`;
  }

  public render (): JSX.Element {
    const closeModal = () => { this.setState({ openModal: false}); };

    return (
      <div className="nabi-container">
        <PageTitle pageTitle="REGISTRATION" />

        <div className="nabi-background-white nabi-section">
          <RegistrationForm
            handleChange={this.handleChange}
            handleSubmit={this.handleSubmit}
            handleBirthdayChange={this.handleBirthdayChange}
            hearAboutUs={this.state.hearAboutUs}
            birthday={this.state.birthday}
            selectedRole={this.state.role}
            fields={this.state.fields}
          />
        </div>
        <AgeDisclaimer
          isFormDialogOpen={this.state.openModal}
          closeHandler={closeModal}
        />
        {this.state.fireRedirect && this.state.role === 'instructor' && (
          <Redirect to={`welcome-instructor/${this.state.id}`} />
        )}
      </div>
    );
  }
}

export default Registration;
