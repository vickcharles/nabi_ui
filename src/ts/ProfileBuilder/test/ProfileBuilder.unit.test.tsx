import * as React from 'react';
import { ProfileBuilder } from '../';
import { shallow } from 'enzyme';
import { Role, UserState } from '../../Registration/model';

describe('ProfileBuilder', () => {
  let wrapper: any;
  let classes: object;
  let theme: object;
  let changeAvatar: any;

  const mockUser: UserState = {
    id: '',
    firstName: 'foo',
    lastName: 'bar',
    email: 'oreo',
    password: '',
    zipCode: '',
    role: Role.instructor,
    hearAboutUs: ''
  };

  beforeEach(() => {
    changeAvatar = (id: string, avatar: string): void => { console.log( avatar ); };
    wrapper = shallow(
      <ProfileBuilder
        classes={classes}
        theme={theme}
        users={[mockUser]}
        changeAvatar={changeAvatar}
      />
    );
  });

  it('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
