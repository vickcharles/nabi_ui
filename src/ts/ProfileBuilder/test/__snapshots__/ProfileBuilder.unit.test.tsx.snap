// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileBuilder Matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withRouter(WithStyles(ProfileBuilder))
    classes={undefined}
    theme={undefined}
    users={
      Array [
        Object {
          "email": "oreo",
          "firstName": "foo",
          "hearAboutUs": "",
          "id": "",
          "lastName": "bar",
          "password": "",
          "role": "instructor",
          "zipCode": "",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "render": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "render": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
